Sources/SwiftCompilerPluginMessageHandling/JSON/JSONEncoding.swift

```diff
@@ -0,0 +1,605 @@
+  }
+
+  @inline(__always)
+  static func string(_ str: String) -> JSONReference {
+    .init(backing: .string(str))
+  }
+  @inline(__always)
+  static func number(_ integer: some BinaryInteger & LosslessStringConvertible) -> JSONReference {
+    .init(backing: .number(String(integer)))
+  }
+  @inline(__always)
+  static func number(_ floating: some BinaryFloatingPoint & LosslessStringConvertible) -> JSONReference {
+    .init(backing: .number(String(floating)))
+  }
+}
+
+/// Serialize JSONReference to [UInt8] data.
+private struct JSONWriter {
+  var data: [UInt8]
+  init() {
+    data = []
+  }
+
+  mutating func write(ascii: UInt8) {
+    data.append(ascii)
+  }
+  mutating func write(scalar: UnicodeScalar) {
+    data.append(UInt8(ascii: scalar))
+  }
+  mutating func write(string: StaticString) {
+    string.withUTF8Buffer { buffer in
+      data.append(contentsOf: buffer)
+    }
+  }
+  mutating func write(utf8: some Collection<UInt8>) {
+    data.append(contentsOf: utf8)
+  }
+
+  mutating func serialize(value: JSONReference) {
+    switch value.backing {
+    case .null:
+      write(string: "null")
+    case .trueKeyword:
+      write(string: "true")
+    case .falseKeyword:
+      write(string: "false")
+    case .string(let string):
+      serialize(string: string)
+    case .number(var string):
+      string.withUTF8 {
+        write(utf8: $0)
+      }
+    case .array(let array):
+      serialize(array: array)
+    case .object(let dictionary):
+      serialize(object: dictionary)
+    }
+  }
+
+  mutating func serialize(string: String) {
+    var string = string
+    write(scalar: "\"")
+    string.withUTF8 { utf8 in
+      var cursor = utf8.baseAddress!
+      let end = utf8.baseAddress! + utf8.count
+      var mark = cursor
+      func flush() {
+        if cursor > mark {
+          write(utf8: UnsafeBufferPointer(start: mark, count: cursor - mark))
+        }
+        cursor += 1
+        mark = cursor
+      }
+      while cursor != end {
+        switch cursor.pointee {
+        case UInt8(ascii: "\""):
+          flush()
+          write(string: "\\\"")
+        case UInt8(ascii: "\\"):
+          flush()
+          write(string: "\\\\")
+        case 0x08:
+          flush()
+          write(string: "\\b")
+        case 0x09:
+          flush()
+          write(string: "\\t")
+        case 0x0A:
+          flush()
+          write(string: "\\n")
+        case 0x0C:
+          flush()
+          write(string: "\\f")
+        case 0x0D:
+          flush()
+          write(string: "\\r")
+        case 0x00...0x1F:
+          let c = cursor.pointee
+          flush()
+          write(string: "\\u00")
+          let _0 = UInt8(ascii: "0")
+          let _A = UInt8(ascii: "A")
+          for shift in stride(from: 4, through: 0, by: -4) {
+            let d = (c >> shift) & 0xF
+            write(ascii: d < 10 ? (_0 + d) : (_A + d - 10))
+          }
+        default:
+          // Accumulate this byte.
```

Comment:
Did you see any mention that JSON strings need to be encoded as UTF-8, which is what we are assuming here? I read through the standard and didn’t see any mention of which encoding JSON is expected to use. 

Probably fine in practice for us though.

Reply:
I don't think JSON spec itself defines it to be UTF-8 encoded. You can have an any JSON text  with any text encoding. But each decoder/encoder implementation usually only support one encoding, and almost all implementations chose UTF-8. Still, you can transcode a UTF-8 JSON text to e.g. UTF-32, and it's still a JSON text.

But, for transferring JSON text, IETF RFC8259 says https://datatracker.ietf.org/doc/html/rfc8259
> JSON text exchanged between systems that are not part of a closed ecosystem MUST be encoded using UTF-8

Reply:
Hah, I guess that does effectively specify the encoding to be UTF-8. Thanks for finding that document.